File: code/ch02/2.7.2.largestNumber.java
import java.util.*;

public class Solution {
  public String largestNumber(int[] nums) {
    String[] s = new String[nums.length];
    for (int i = 0; i < nums.length; i++)
      s[i] = String.valueOf(nums[i]);
    List<String> str = Arrays.asList(s);
    str.sort((String a, String b) -> {
      String left = a + b, right = b + a;
      return right.compareTo(left);
    });
    String ans = "";
    for (int i = 0; i < nums.length; i++)
      ans = ans + str.get(i);
    String real_ans = "";
    boolean flag = false;
    for (int i = 0; i < ans.length(); i++) {
      if (!flag) flag = !(ans.charAt(i) == '0');
      if (flag) real_ans = real_ans + ans.substring(i, i + 1);
    }
    if (!flag) return "0";
    else return real_ans;
  }
}

File: code/ch02/2.7.2.largestNumber.cpp
#include <vector>
#include <algorithm>
#include <cstdio>
#include <string>
#include <cstdlib>
using namespace std;

string largestNumber(const vector<int>& nums) {
  vector<string> s;
  for (int i = 0; i < nums.size(); i++)
    s.push_back(to_string(nums[i]));
  // clang++/g++: need parameter '-std=c++11'
  sort(s.begin(), s.end(), [](string a, string b)
      -> bool {
      return a + b > b + a;
      });
  string ans("");
  for (int i = 0; i < nums.size(); i++)
    ans += s[i];
    string real_ans("");
    bool flag = false;
    for (int i = 0; i < ans.length(); i++) {
      if (!flag) flag = !(ans[i] == '0');
      if (flag) real_ans.push_back(ans[i]);
    }
    if (!flag) return string("0");
    else return real_ans;
}
